keps/sig-auth/3299-kms-v2-improvements/README.md

```diff
@@ -242,34 +243,48 @@ message StatusResponse {
- 
-+    CommonKeyID *string `json:"commonKeyID,omitempty" protobuf:"bytes,4,opt,name=commonKeyID"`
-+
-	 // The latest available observations of the storageVersion's state.
-	 // +optional
-	 // +listType=map
-@@ -77,6 +79,8 @@ type ServerStorageVersion struct {
-	 // The encodingVersion must be included in the decodableVersions.
-	 // +listType=set
-	 DecodableVersions []string `json:"decodableVersions,omitempty" protobuf:"bytes,3,opt,name=decodableVersions"`
-+
-+    KeyID *string `json:"keyID,omitempty" protobuf:"bytes,4,opt,name=keyID"`
- }
- 
- type StorageVersionConditionType string
-```
+The `key_id` will be funneled into the status of the new encryption state REST API.  Each API server will create this resource with a random name on startup (stale objects will be garbage collected by a new controller).  This resource provides the necessary information for a controller to automatically initiate storage migration when consensus has been reached among the different API server instances.  Opaque hashes are used to prevent leaking any specific information about the encryption configuration and the key IDs in use.  This is a cluster scoped REST API that is only meant to be read by the associated controllers and the cluster admin.
+
+```go
+// metadata.name is the ID of the API server
+type EncryptionState struct {
+    metav1.TypeMeta `json:",inline"`
+    // +optional
+    metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+    Spec EncryptionStateSpec `json:"spec" protobuf:"bytes,2,opt,name=spec"`
 
-> NOTE: Since the storage version API is still alpha, this KEP will simply aim to make it possible to have automated rotation when that API is enabled and has been updated to include the new fields. The rotation feature will first be scoped to a single API server and will not be part of the graduation criteria for this KEP.
+    Status EncryptionStateStatus `json:"status" protobuf:"bytes,3,opt,name=status"`
+}
+
+// spec is empty
+type EncryptionStateSpec struct{}
+
+type EncryptionStateStatus struct {
+    EncryptionConfigurationHash Hash `json:"encryptionConfigurationHash" protobuf:"bytes,1,opt,name=encryptionConfigurationHash"`
+
+    Resources []EncryptionStateResource `json:"resources" protobuf:"bytes,2,opt,name=resources"`
+}
+
+type EncryptionStateResource struct {
+    Resource metav1.GroupResource `json:"resource" protobuf:"bytes,1,opt,name=resource"`
+
+    WriteKeyIDHash Hash `json:"writeKeyIDHash" protobuf:"bytes,2,opt,name=writeKeyIDHash"`
+
+    ReadKeyIDHashes []Hash `json:"readKeyIDHashes" protobuf:"bytes,3,opt,name=readKeyIDHashes"`
+
+    // this field would be populated by scanning all KMS v2 encrypted data and recording all observed
+    // key IDs.  this would be an expensive operation and is a bit weird to be per API server.
+    InUseKeyIDHashes []Hash `json:"inUseKeyIDHashes" protobuf:"bytes,4,opt,name=inUseKeyIDHashes"`
```

Comment:
Reminder: this needs a way to indicate that non-kms v2 encryption methods are in use as well as when unencrypted data is present.